### ISSUE TITLE
Sidekiq queues are misused

### ISSUE THREAD

**Issue by ClearlyClaire** [mastodon/mastodon#8342]
Sidekiq queues are misused.
Indeed, unless queues are given weight, sidekiq pulls jobs from the queues in the order they are defined.
This means that, if you run sidekiq according to the systemd file defined in https://github.com/tootsuite/documentation/blob/master/Running-Mastodon/Production-guide.md, no “mailers” job will run before all “default” ones are processed, no “pull” will be processed before “default” or “mailers” etc.
I don't think they were defined to imply any priority, but it is what they currently do.
In addition, the systemd unit file referenced above do not agree with the order defined in `config/sidekiq.yml`.

I believe there is no reason to not have just three queues, as `critical`, `default`, `low`, with most of the stuff in `default`, and some maintenance jobs like `remote_remove` in low.

- Sidekiq documentation on queues and priority: https://github.com/mperham/sidekiq/wiki/Advanced-Options#queues
- Some tips on using them: https://philsturgeon.uk/2016/11/16/tips-on-sidekiq-queues/

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by Gargron** [comment]
I'm against this for the following reasons:

When you scale beyond a single Sidekiq process, you usually split by queue. So I'll have a 5 threads process for mailers, two 45 threads processes for default, same for push, etc. It's very very useful because you can make sure that certain jobs aren't held up by other jobs, e.g. the push queue could be lagging behind but local users would be fine if the default queue is okay.

Secondly the reason I have not changed or renamed the queues so far is that it would require all admins to go into their systemd files and whatnot and change it around. It's a big headache but even more than that, historically people have glossed over changes like that, which results in a lot of troubleshooting issues on this bug tracker and broken servers.

**Comment by Gargron** [comment]
TL;DR: The categorization of jobs is currently pretty good, I would not move push jobs into the default queue at all. Renaming them to be more neutral, i.e. default to critical, push to high, mailers to normal, pull to low, is a noble goal, but probably not possible. I don't think your migration plan is viable because I remember the reason 4 separate queues were needed in the first place, there's no way a release would survive with just 2.

**Comment by Gargron** [comment]
The push queue is high-priority because you'd want your followers on other servers to get your new posts, or your update profile, or the fact that you deleted something. The pull queue, you can live if it doesn't work at all. It contains things that could run slowly (re-fetching profile, resolving an entire thread of a conversation, link preview for some slow website).

**Comment by ThibG** [comment]
Ah, alright. But it's *not* what the default configuration says (docker or the one linked in the production guide), in term of priorities. We should at least fix that.

**Comment by Gargron** [comment]
How about we add weights to the default config, with the queues we already have? default:6, push:4, mailers:2, pull:1

**Comment by ThibG** [comment]
This means something like the maintenance operations will slow down the whole thing (well, they already halt outbound federation with the default config due to misordered queues), which I'm not too sure about, but… yeah why not

**Comment by ThibG** [comment]
Since it's mostly a configuration issue that probably won't be picked up by admins if we update the docs anyway, wouldn't it make sense to notify them on the admin mailing-list about this issue, and let them experiment with what works for them, suggesting the `-q default,6 -q push,4 -q mailers,2 -q pull,1` default?

**Comment by unarist** [comment]
> e.g. the push queue could be lagging behind but local users would be fine if the default queue is okay.

I also want to remind about ProcessingWorker. How many ProcessingWorker jobs enqueued will heavily depend to *outside* of the instance. Remote instance may send bunch of deletion, reblog of posts in dead server, bunch of follows due to imports, etc. We may want to limit incoming activities before queueing, though.

**Comment by ThibG** [comment]
Alright, I opened a PR (#8345) to fix a few things related to workers and queues according to my understanding. I don't think we need weighted queues with that config, but we need to advertise the priority order of queues, in the documentation, on the mailing list, and maybe in the release notes for good measures.

**Comment by trwnh** [comment]
i think the sidekiq queues have changed a few times since this issue was filed, so it bears fresh exploration in a separate issue
